lib/src/git_backend.rs

```diff
@@ -445,6 +445,28 @@ impl Backend for GitBackend {
     }
 
     fn write_commit(&self, contents: &Commit) -> BackendResult<CommitId> {
+        let mut mut_commit = contents.clone();
+        // It's possible a commit already exists with the same commit id but different
+        // change id. Adjust the timestamp until this is no longer the case.
+        loop {
+            match self.write_commit_internal(&mut_commit)? {
```

Comment:
Much of `write_commit_internal()` could be moved out of it and into this function (we shouldn't have to write the tree on every iteration, for example), but I don't think it's worth it since this happens so rarely.

Reply:
That was obviously supposed to be a reply to your comment about underflow when subtracting from the timestamp

Reply:
> Much of write_commit_internal() could be moved out of it and into this function

I don't see many advantages of doing that. I think that if we tried really hard, we could save a single clone of a Commit (which should be quite fast). 

> We actually allow negative timestamps

That's neat. But why? Is someone somewhere working on turning the process of writing the Gutenberg Bible into a git repository?

Reply:
> > Much of write_commit_internal() could be moved out of it and into this function
> 
> I don't see many advantages of doing that. I think that if we tried really hard, we could save a single clone of a Commit (which should be quite fast).

We would avoid writing the tree object multiple times, right? Again, I think it's rare enough that the retries happen that we shouldn't bother.

> 
> > We actually allow negative timestamps
> 
> That's neat. But why? Is someone somewhere working on turning the process of writing the Gutenberg Bible into a git repository?

The link in my commit message is stale. https://github.com/git/git.github.io/blob/master/_posts/2022-09-31-edition-91.markdown is where it's now. The example mentioned there was "Apollo code".

Reply:
There's a joke there somewhere about rebasing Apollo code on top of the Space Shuttle code to get the code for the SLS.

Reply:
> We would avoid writing the tree object multiple times, right?

I misread the code; we don't write trees in `write_commit()`. We could avoid *reading* the tree, though.